### PR TITLE
Fixes #204 Independent sub modules for codecs ilp, btp, ildcp

### LINKED ISSUE #204: Consider a `codecs` sub-module that holds all codecs.

### PR DESCRIPTION
**Pull request by sudheesh001**
- Creates `codecs` aggregated `pom` to hold all `codecs` for (`btp`, `ildcp`, `ilp`)
- Moves `codec-framework` to `codecs` module.
- Additionally Creates individual `codecs-btp`, `codecs-ildcp` and `codecs-ilp` modules.
- Refactors current core codec modules to the newer modules.
- Reduces the number of dependencies required for some of the modules.
- Moves corresponding tests for each of the codec modules.

Signed-off-by: sudheesh001 <sudheesh1995@outlook.com>

### CODE REVIEW

**Comment by sudheesh001** [comment]
@sappenin Please review this patch when you're free. I'll be spending some time reading through ILP in more detail before working on the other patches. 😄

File: codecs/codecs-btp/src/main/java/org/interledger/codecs/btp/AsnBtpPacketCodec.java [hunk 1]
```diff
@@ -71,7 +76,7 @@ public void encode(T value) {
 
   protected void onRequestIdChanged(long code) {
 
-    BtpMessageType type = BtpMessageType.fromCode(getValueAt(0));
+    BtpMessageType type = fromCode(getValueAt(0));
```
Comment (sappenin) [inline codecs/codecs-btp/src/main/java/org/interledger/codecs/btp/AsnBtpPacketCodec.java]: Minor nit: I prefer to keep method imports class-qualified, like this: `BtpMessageType.fromCode(getValueAt(0));` so it's overt where this is coming from. It's probably not a big deal in this case, but I have had Intellij auto-qualify static methods and if there's ever ambiguity it sometimes picks the wrong one (and it's easier to miss if the actual class is statically imported at the top of the file).

The one exception to this style rule is in unit-test assertions, where I think readability outweighs things.

File: codecs/codecs-btp/src/test/java/org/interledger/codecs/btp/AsnBtpErrorDataSerializationTest.java [hunk 2]
```diff
@@ -147,10 +152,10 @@ public void read() throws Exception {
      */
     final BtpError actualValue = codecContext.read(BtpError.class, inputStream);
 
-    assertThat(actualValue.getRequestId(), is(this.expectedRequestId));
-    assertThat(actualValue.getErrorCode(), is(this.expectedErrorCode));
-    assertThat(actualValue.getTriggeredAt(), is(this.expectedTriggeredAt));
-    assertThat(actualValue.getErrorData(), is(BaseEncoding.base64().decode(this.expectedErrorDataBase64Bytes)));
+    MatcherAssert.assertThat(actualValue.getRequestId(), Matchers.is(this.expectedRequestId));
```
Comment (sappenin) [inline codecs/codecs-btp/src/test/java/org/interledger/codecs/btp/AsnBtpErrorDataSerializationTest.java]: Minor nit: Let's go back to just `assertThat(...`. In the case of unit tests, IMHO, using `MatcherAssert.assertThat` everywhere impedes readability without much gain.

File: codecs/codecs-ildcp/src/test/java/org/interledger/codecs/ildcp/AsnIldcpResponseCodecTest.java [hunk 3]
```diff
@@ -59,28 +60,28 @@ public void setUp() {
   @Test
   public void decode() {
     codec.encode(RESPONSE);
-    assertThat(codec.getCodecAt(0).decode(), is(FOO_ADDRESS));
-    assertThat(codec.getCodecAt(1).decode(), is((short) 9));
-    assertThat(codec.getCodecAt(2).decode(), is(BTC));
+    assertThat(codec.getCodecAt(0).decode(), Is.is(FOO_ADDRESS));
```
Comment (sappenin) [inline codecs/codecs-ildcp/src/test/java/org/interledger/codecs/ildcp/AsnIldcpResponseCodecTest.java]: Per my comments above...

```suggestion
    assertThat(codec.getCodecAt(0).decode(), is(FOO_ADDRESS));
```

File: pom.xml [hunk 4]
```diff
@@ -100,12 +100,10 @@
   <modules>
     <module>codec-framework</module>
```
Comment (sappenin) [inline pom.xml]: I think we should move this module (`codec-framework`) into `codecs`.

File: codecs/codecs-btp/src/main/java/org/interledger/codecs/btp/framework/BtpCodecContextFactory.java [hunk 5]
```diff
@@ -1,4 +1,4 @@
-package org.interledger.btp.asn.framework;
+package org.interledger.codecs.btp.framework;
```
Comment (sappenin) [inline codecs/codecs-btp/src/main/java/org/interledger/codecs/btp/framework/BtpCodecContextFactory.java]: With the new structure introduced by this PR, I think this package should just be `org.interledger.codecs.btp`. The usage of the extra package named `framework` in each module just for a single class doesn't seem to add any value.

```suggestion
package org.interledger.codecs.btp;
```

File: codecs/codecs-ildcp/src/main/java/org/interledger/codecs/ildcp/framework/IldcpCodecContextFactory.java [hunk 6]
```diff
@@ -1,4 +1,4 @@
-package org.interledger.ildcp.asn.framework;
+package org.interledger.codecs.ildcp.framework;
```
Comment (sappenin) [inline codecs/codecs-ildcp/src/main/java/org/interledger/codecs/ildcp/framework/IldcpCodecContextFactory.java]:
```suggestion
package org.interledger.codecs.ildcp;
```

File: codecs/codecs-ilp/src/main/java/org/interledger/codecs/ilp/framework/InterledgerCodecContextFactory.java [hunk 7]
```diff
@@ -1,4 +1,4 @@
-package org.interledger.core.asn.framework;
+package org.interledger.codecs.ilp.framework;
```
Comment (sappenin) [inline codecs/codecs-ilp/src/main/java/org/interledger/codecs/ilp/framework/InterledgerCodecContextFactory.java]:
```suggestion
package org.interledger.codecs.ilp;
```

**Comment by sudheesh001** [comment]
Thanks @sappenin . Made the changes and correct usages of `Matchers.assertThat` throughout along with `Is`. Also moved `codec-framework` to `codecs` as suggested. Please review 😄

File: codecs/codecs-btp/pom.xml [hunk 8]
```diff
@@ -9,46 +9,30 @@
...
-      <groupId>com.google.guava</groupId>
-      <artifactId>guava</artifactId>
-      <scope>test</scope>
+      <groupId>org.interledger</groupId>
+      <artifactId>btp-core</artifactId>
     </dependency>
-
   </dependencies>
 
+  <properties>
```
Comment (sappenin) [inline codecs/codecs-btp/pom.xml]: This should be necessary - it should be defined in the parent pom and inherited.

File: codecs/codecs-ildcp/src/test/java/org/interledger/codecs/ildcp/AsnIldcpPacketCodecTest.java [hunk 9]
```diff
@@ -60,33 +60,33 @@ public void setup() {
   public void encodeDecode() {
     final IldcpRequestPacket packet = IldcpRequestPacket.builder().build();
     codec.encode(packet);
-    assertThat(codec.getCodecAt(0).decode(), is((short) 12));
-    assertThat(codec.getCodecAt(1).decode(), is(packet));
+    assertThat(codec.getCodecAt(0).decode(), CoreMatchers.is((short) 12));
```
Comment (sappenin) [inline codecs/codecs-ildcp/src/test/java/org/interledger/codecs/ildcp/AsnIldcpPacketCodecTest.java]: Can we remove `CoreMatchers` everywhere?

File: codecs/codecs-ilp/src/main/java/org/interledger/codecs/ilp/InterledgerCodecContextFactory.java [hunk 10]
```diff
@@ -29,16 +40,7 @@
 import org.interledger.core.InterledgerPacket;
 import org.interledger.core.InterledgerPreparePacket;
 import org.interledger.core.InterledgerRejectPacket;
-import org.interledger.core.asn.codecs.AsnConditionCodec;
```
Comment (sappenin) [inline codecs/codecs-ilp/src/main/java/org/interledger/codecs/ilp/InterledgerCodecContextFactory.java]: Does this not break checkstyle? If memory serves there's an overt ordering of imports. I can check this out a bit later...

File: codecs/codecs-ilp/src/test/java/org/interledger/codecs/ilp/InterledgerRejectPacketOerSerializerTests.java [hunk 11]
```diff
@@ -181,12 +182,12 @@ public void testIndividualRead() throws IOException {
 
     final ByteArrayInputStream asn1OerErrorBytesStream = constructInterledgerRejectPacketAsn1OerInputStream();
     final InterledgerRejectPacket error = context.read(InterledgerRejectPacket.class, asn1OerErrorBytesStream);
-    assertThat(error, is(packet));
+    assertThat(error, CoreMatchers.is(packet));
```
Comment (sappenin) [inline codecs/codecs-ilp/src/test/java/org/interledger/codecs/ilp/InterledgerRejectPacketOerSerializerTests.java]:
```suggestion
    assertThat(error, is(packet));
```

File: codecs/codecs-ildcp/src/test/java/org/interledger/codecs/ildcp/AsnIldcpResponseCodecTest.java (hunk 3 above)
Comment (sappenin) [reply codecs/codecs-ildcp/src/test/java/org/interledger/codecs/ildcp/AsnIldcpResponseCodecTest.java]: I think the only remaining thing here is to remove all the `CoreMatchers` prefixes in the unit tests.

File: codecs/codecs-btp/src/main/java/org/interledger/codecs/btp/framework/BtpCodecContextFactory.java (hunk 5 above)
Comment (sappenin) [reply codecs/codecs-btp/src/main/java/org/interledger/codecs/btp/framework/BtpCodecContextFactory.java]: LGTM!

File: codecs/codecs-ilp/src/main/java/org/interledger/codecs/ilp/InterledgerCodecContextFactory.java (hunk 10 above)
Comment (sappenin) [reply codecs/codecs-ilp/src/main/java/org/interledger/codecs/ilp/InterledgerCodecContextFactory.java]: So, looks like checkstyle isn't really running during the build. I created #224 for tracking. If you want to push this off to there, that's fine.

Comment (sudheesh001) [reply codecs/codecs-ilp/src/main/java/org/interledger/codecs/ilp/InterledgerCodecContextFactory.java]: Yes, It seems to work locally and gives me a bunch of errors when I try to build using `mvn clean install` regarding the lexographical ordering but once I fix it it passes. I can investigate it further once this patch gets merged.

File: codecs/codecs-ildcp/src/test/java/org/interledger/codecs/ildcp/AsnIldcpResponseCodecTest.java (hunk 3 above)
Comment (sudheesh001) [reply codecs/codecs-ildcp/src/test/java/org/interledger/codecs/ildcp/AsnIldcpResponseCodecTest.java]: removed the `CoreMatchers` in the latest commit.

File: codecs/codecs-ilp/src/main/java/org/interledger/codecs/ilp/InterledgerCodecContextFactory.java (hunk 10 above)
Comment (sappenin) [reply codecs/codecs-ilp/src/main/java/org/interledger/codecs/ilp/InterledgerCodecContextFactory.java]: Sounds good.

**Pull request merged**